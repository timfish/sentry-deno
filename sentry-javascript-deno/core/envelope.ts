// deno-lint-ignore-file 
import {
  Baggage,
  DsnComponents,
  DynamicSamplingContext,
  Event,
  EventEnvelope,
  EventEnvelopeHeaders,
  EventItem,
  SdkInfo,
  SdkMetadata,
  Session,
  SessionAggregates,
  SessionEnvelope,
  SessionItem,
} from '../types/mod.ts';
import { createEnvelope, dropUndefinedKeys, dsnToString, getSentryBaggageItems } from '../utils/mod.ts';

/** Extract sdk info from from the API metadata */
function getSdkMetadataForEnvelopeHeader(metadata?: SdkMetadata): SdkInfo | undefined {
  if (!metadata || !metadata.sdk) {
    return;
  }
  const { name, version } = metadata.sdk;
  return { name, version };
}

/**
 * Apply SdkInfo (name, version, packages, integrations) to the corresponding event key.
 * Merge with existing data if any.
 **/
function enhanceEventWithSdkInfo(event: Event, sdkInfo?: SdkInfo): Event {
  if (!sdkInfo) {
    return event;
  }
  event.sdk = event.sdk || {};
  event.sdk.name = event.sdk.name || sdkInfo.name;
  event.sdk.version = event.sdk.version || sdkInfo.version;
  event.sdk.integrations = [...(event.sdk.integrations || []), ...(sdkInfo.integrations || [])];
  event.sdk.packages = [...(event.sdk.packages || []), ...(sdkInfo.packages || [])];
  return event;
}

/** Creates an envelope from a Session */
export function createSessionEnvelope(
  session: Session | SessionAggregates,
  dsn: DsnComponents,
  metadata?: SdkMetadata,
  tunnel?: string,
): SessionEnvelope {
  const sdkInfo = getSdkMetadataForEnvelopeHeader(metadata);
  const envelopeHeaders = {
    sent_at: new Date().toISOString(),
    ...(sdkInfo && { sdk: sdkInfo }),
    ...(!!tunnel && { dsn: dsnToString(dsn) }),
  };

  const envelopeItem: SessionItem =
    'aggregates' in session ? [{ type: 'sessions' }, session] : [{ type: 'session' }, session];

  return createEnvelope<SessionEnvelope>(envelopeHeaders, [envelopeItem]);
}

/**
 * Create an Envelope from an event.
 */
export function createEventEnvelope(
  event: Event,
  dsn: DsnComponents,
  metadata?: SdkMetadata,
  tunnel?: string,
): EventEnvelope {
  const sdkInfo = getSdkMetadataForEnvelopeHeader(metadata);
  const eventType = event.type || 'event';

  const { transactionSampling } = event.sdkProcessingMetadata || {};
  const { method: samplingMethod, rate: sampleRate } = transactionSampling || {};

  enhanceEventWithSdkInfo(event, metadata && metadata.sdk);

  const envelopeHeaders = createEventEnvelopeHeaders(event, sdkInfo, tunnel, dsn);

  // Prevent this data (which, if it exists, was used in earlier steps in the processing pipeline) from being sent to
  // sentry. (Note: Our use of this property comes and goes with whatever we might be debugging, whatever hacks we may
  // have temporarily added, etc. Even if we don't happen to be using it at some point in the future, let's not get rid
  // of this `delete`, lest we miss putting it back in the next time the property is in use.)
  delete event.sdkProcessingMetadata;

  const eventItem: EventItem = [
    {
      type: eventType,
      sample_rates: [{ id: samplingMethod, rate: sampleRate }],
    },
    event,
  ];
  return createEnvelope<EventEnvelope>(envelopeHeaders, [eventItem]);
}

function createEventEnvelopeHeaders(
  event: Event,
  sdkInfo: SdkInfo | undefined,
  tunnel: string | undefined,
  dsn: DsnComponents,
): EventEnvelopeHeaders {
  const baggage: Baggage | undefined = event.sdkProcessingMetadata && event.sdkProcessingMetadata.baggage;
  const dynamicSamplingContext = baggage && getSentryBaggageItems(baggage);

  return {
    event_id: event.event_id as string,
    sent_at: new Date().toISOString(),
    ...(sdkInfo && { sdk: sdkInfo }),
    ...(!!tunnel && { dsn: dsnToString(dsn) }),
    ...(event.type === 'transaction' &&
      dynamicSamplingContext && {
        trace: dropUndefinedKeys({ ...dynamicSamplingContext }) as DynamicSamplingContext,
      }),
  };
}
