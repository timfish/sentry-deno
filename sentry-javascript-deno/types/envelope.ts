// deno-lint-ignore-file 
import { ClientReport } from './clientreport.ts';
import { DsnComponents } from './dsn.ts';
import { Event } from './event.ts';
import { SdkInfo } from './sdkinfo.ts';
import { Session, SessionAggregates } from './session.ts';
import { Transaction, TransactionSamplingMethod } from './transaction.ts';
import { UserFeedback } from './user.ts';

// Based on: https://develop.sentry.dev/sdk/envelopes/

// Based on https://github.com/getsentry/relay/blob/b23b8d3b2360a54aaa4d19ecae0231201f31df5e/relay-sampling/src/lib.rs#L685-L707
export type DynamicSamplingContext = {
  trace_id: Transaction['traceId'];
  public_key: DsnComponents['publicKey'];
  sample_rate?: string;
  release?: string;
  environment?: string;
  transaction?: string;
  user_segment?: string;
};

export type EnvelopeItemType =
  | 'client_report'
  | 'user_report'
  | 'session'
  | 'sessions'
  | 'transaction'
  | 'attachment'
  | 'event';

export type BaseEnvelopeHeaders = {
  [key: string]: unknown;
  dsn?: string;
  sdk?: SdkInfo;
};

export type BaseEnvelopeItemHeaders = {
  [key: string]: unknown;
  type: EnvelopeItemType;
  length?: number;
};

type BaseEnvelopeItem<IH extends BaseEnvelopeItemHeaders, P extends unknown> = [IH, P]; // P is for payload

type BaseEnvelope<EH extends BaseEnvelopeHeaders, I extends BaseEnvelopeItem<BaseEnvelopeItemHeaders, unknown>> = [
  EH,
  I[],
];

type EventItemHeaders = {
  type: 'event' | 'transaction';
  sample_rates?: [{ id?: TransactionSamplingMethod; rate?: number }];
};
type AttachmentItemHeaders = {
  type: 'attachment';
  length: number;
  filename: string;
  content_type?: string;
  attachment_type?: string;
};
type UserFeedbackItemHeaders = { type: 'user_report' };
type SessionItemHeaders = { type: 'session' };
type SessionAggregatesItemHeaders = { type: 'sessions' };
type ClientReportItemHeaders = { type: 'client_report' };

export type EventItem = BaseEnvelopeItem<EventItemHeaders, Event>;
export type AttachmentItem = BaseEnvelopeItem<AttachmentItemHeaders, string | Uint8Array>;
export type UserFeedbackItem = BaseEnvelopeItem<UserFeedbackItemHeaders, UserFeedback>;
export type SessionItem =
  | BaseEnvelopeItem<SessionItemHeaders, Session>
  | BaseEnvelopeItem<SessionAggregatesItemHeaders, SessionAggregates>;
export type ClientReportItem = BaseEnvelopeItem<ClientReportItemHeaders, ClientReport>;

export type EventEnvelopeHeaders = { event_id: string; sent_at: string; trace?: DynamicSamplingContext };
type SessionEnvelopeHeaders = { sent_at: string };
type ClientReportEnvelopeHeaders = BaseEnvelopeHeaders;

export type EventEnvelope = BaseEnvelope<EventEnvelopeHeaders, EventItem | AttachmentItem | UserFeedbackItem>;
export type SessionEnvelope = BaseEnvelope<SessionEnvelopeHeaders, SessionItem>;
export type ClientReportEnvelope = BaseEnvelope<ClientReportEnvelopeHeaders, ClientReportItem>;

export type Envelope = EventEnvelope | SessionEnvelope | ClientReportEnvelope;
export type EnvelopeItem = Envelope[1][number];
