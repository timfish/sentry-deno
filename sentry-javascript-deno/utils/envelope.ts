// deno-lint-ignore-file
import {
  Attachment,
  AttachmentItem,
  DataCategory,
  Envelope,
  EnvelopeItem,
  EnvelopeItemType,
  TextEncoderInternal,
} from '../types/mod.ts';

import { dropUndefinedKeys } from './object.ts';

/**
 * Creates an envelope.
 * Make sure to always explicitly provide the generic to this function
 * so that the envelope types resolve correctly.
 */
export function createEnvelope<E extends Envelope>(
  headers: E[0],
  items: E[1] = [],
): E {
  return [headers, items] as E;
}

/**
 * Add an item to an envelope.
 * Make sure to always explicitly provide the generic to this function
 * so that the envelope types resolve correctly.
 */
export function addItemToEnvelope<E extends Envelope>(
  envelope: E,
  newItem: E[1][number],
): E {
  const [headers, items] = envelope;
  return [headers, [...items, newItem]] as E;
}

/**
 * Convenience function to loop through the items and item types of an envelope.
 * (This function was mostly created because working with envelope types is painful at the moment)
 */
export function forEachEnvelopeItem<E extends Envelope>(
  envelope: Envelope,
  callback: (
    envelopeItem: E[1][number],
    envelopeItemType: E[1][number][0]['type'],
  ) => void,
): void {
  const envelopeItems = envelope[1];
  envelopeItems.forEach((envelopeItem: EnvelopeItem) => {
    const envelopeItemType = envelopeItem[0].type;
    callback(envelopeItem, envelopeItemType);
  });
}

function encodeUTF8(
  input: string,
  textEncoder?: TextEncoderInternal,
): Uint8Array {
  const utf8 = textEncoder || new TextEncoder();
  return utf8.encode(input);
}

/**
 * Serializes an envelope.
 */
export function serializeEnvelope(
  envelope: Envelope,
  textEncoder?: TextEncoderInternal,
): string | Uint8Array {
  const [envHeaders, items] = envelope;

  // Initially we construct our envelope as a string and only convert to binary chunks if we encounter binary data
  let parts: string | Uint8Array[] = JSON.stringify(envHeaders);

  function append(next: string | Uint8Array): void {
    if (typeof parts === 'string') {
      parts = typeof next === 'string'
        ? parts + next
        : [encodeUTF8(parts, textEncoder), next];
    } else {
      parts.push(
        typeof next === 'string' ? encodeUTF8(next, textEncoder) : next,
      );
    }
  }

  for (const item of items) {
    const [itemHeaders, payload] = item as typeof items[number];
    append(`\n${JSON.stringify(itemHeaders)}\n`);
    append(
      typeof payload === 'string' || payload instanceof Uint8Array
        ? payload
        : JSON.stringify(payload),
    );
  }

  return typeof parts === 'string' ? parts : concatBuffers(parts);
}

function concatBuffers(buffers: Uint8Array[]): Uint8Array {
  const totalLength = buffers.reduce((acc, buf) => acc + buf.length, 0);

  const merged = new Uint8Array(totalLength);
  let offset = 0;
  for (const buffer of buffers) {
    merged.set(buffer, offset);
    offset += buffer.length;
  }

  return merged;
}

/**
 * Creates attachment envelope items
 */
export function createAttachmentEnvelopeItem(
  attachment: Attachment,
  textEncoder?: TextEncoderInternal,
): AttachmentItem {
  const buffer = typeof attachment.data === 'string'
    ? encodeUTF8(attachment.data, textEncoder)
    : attachment.data;

  return [
    dropUndefinedKeys({
      type: 'attachment',
      length: buffer.length,
      filename: attachment.filename,
      content_type: attachment.contentType,
      attachment_type: attachment.attachmentType,
    }),
    buffer,
  ];
}

const ITEM_TYPE_TO_DATA_CATEGORY_MAP: Record<EnvelopeItemType, DataCategory> = {
  session: 'session',
  sessions: 'session',
  attachment: 'attachment',
  transaction: 'transaction',
  event: 'error',
  client_report: 'internal',
  user_report: 'default',
};

/**
 * Maps the type of an envelope item to a data category.
 */
export function envelopeItemTypeToDataCategory(
  type: EnvelopeItemType,
): DataCategory {
  return ITEM_TYPE_TO_DATA_CATEGORY_MAP[type];
}
