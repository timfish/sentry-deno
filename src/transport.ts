import { createTransport } from '../sentry-javascript-deno/core/mod.ts';
import {
  BaseTransportOptions,
  Transport,
  TransportMakeRequestResponse,
  TransportRequest,
} from '../sentry-javascript-deno/types/mod.ts';

export interface DenoTransportOptions extends BaseTransportOptions {
  headers?: Record<string, string>;
  fetchOptions?: RequestInit;
}

/**
 * Creates a Transport that uses the Fetch API to send events to Sentry.
 */
export function makeFetchTransport(options: DenoTransportOptions): Transport {
  async function makeRequest(
    request: TransportRequest,
  ): Promise<TransportMakeRequestResponse> {
    // Check if we have permissions to send this event
    const url = new URL(options.url);

    const permission = await Deno.permissions.query({
      name: 'net',
      host: url.host,
    });

    if (permission.state !== 'granted') {
      console.warn(
        'Event was not sent due to missing permissions. Run Deno with --allow-net to allow sending of events.',
      );

      return {};
    }

    const requestOptions: RequestInit = {
      body: request.body,
      method: 'POST',
      referrerPolicy: 'origin',
      headers: options.headers,
      ...options.fetchOptions,
    };

    return fetch(options.url, requestOptions).then((response) => ({
      statusCode: response.status,
      headers: {
        'x-sentry-rate-limits': response.headers.get('X-Sentry-Rate-Limits'),
        'retry-after': response.headers.get('Retry-After'),
      },
    }));
  }

  return createTransport(options, makeRequest);
}
