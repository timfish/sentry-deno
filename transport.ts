import {
  createTransport,
  BaseTransportOptions,
  Transport,
  TransportRequest,
  TransportMakeRequestResponse,
} from "./deps.ts";

export interface DenoTransportOptions extends BaseTransportOptions {
  headers?: Record<string, string>;
  fetchOptions?: RequestInit;
}

/**
 * Creates a Transport that uses the Fetch API to send events to Sentry.
 */
export function makeFetchTransport(options: DenoTransportOptions): Transport {
  function makeRequest(
    request: TransportRequest
  ): PromiseLike<TransportMakeRequestResponse> {
    const requestOptions: RequestInit = {
      body: request.body,
      method: "POST",
      referrerPolicy: "origin",
      headers: options.headers,
      ...options.fetchOptions,
    };

    return fetch(options.url, requestOptions).then((response) => ({
      statusCode: response.status,
      headers: {
        "x-sentry-rate-limits": response.headers.get("X-Sentry-Rate-Limits"),
        "retry-after": response.headers.get("Retry-After"),
      },
    }));
  }

  return createTransport(options, makeRequest);
}
