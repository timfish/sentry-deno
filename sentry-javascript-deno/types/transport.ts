// deno-lint-ignore-file
import { EventDropReason } from './clientreport.ts';
import { DataCategory } from './datacategory.ts';
import { Envelope } from './envelope.ts';
import { TextEncoderInternal } from './textencoder.ts';

export type TransportRequest = {
  body: string | Uint8Array;
};

export type TransportMakeRequestResponse = {
  statusCode?: number;
  headers?: {
    [key: string]: string | null;
    'x-sentry-rate-limits': string | null;
    'retry-after': string | null;
  };
};

export interface InternalBaseTransportOptions {
  bufferSize?: number;
  recordDroppedEvent: (
    reason: EventDropReason,
    dataCategory: DataCategory,
  ) => void;
  textEncoder?: TextEncoderInternal;
}

export interface BaseTransportOptions extends InternalBaseTransportOptions {
  // url to send the event
  // transport does not care about dsn specific - client should take care of
  // parsing and figuring that out
  url: string;
}

export interface Transport {
  send(request: Envelope): PromiseLike<void>;
  flush(timeout?: number): PromiseLike<boolean>;
}

export type TransportRequestExecutor = (
  request: TransportRequest,
) => PromiseLike<TransportMakeRequestResponse>;
