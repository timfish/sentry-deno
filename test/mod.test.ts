import { defaultIntegrations, DenoClient } from '../mod.ts';
import { makeTestTransport } from './transport.ts';
import {
  createStackParser,
  nodeStackLineParser,
} from '../sentry-javascript-deno/utils/mod.ts';
import { Event, Integration } from '../sentry-javascript-deno/types/mod.ts';
import { Hub, Scope } from '../sentry-javascript-deno/hub/mod.ts';
import { getNormalizedEvent } from './envelope.ts';
import { assertSnapshot } from 'https://deno.land/std@0.151.0/testing/snapshot.ts';

function getTestClient(
  callback: (event?: Event) => void,
  integrations: Integration[] = [],
): [Hub, DenoClient] {
  const client = new DenoClient({
    dsn: 'https://233a45e5efe34c47a3536797ce15dafa@nothing.here/5650507',
    debug: true,
    integrations: [...defaultIntegrations, ...integrations],
    stackParser: createStackParser(nodeStackLineParser()),
    transport: makeTestTransport((envelope) => {
      callback(getNormalizedEvent(envelope));
    }),
  });

  const scope = new Scope();
  const hub = new Hub(client, scope);

  return [hub, client];
}

Deno.test('captureException', (t) => {
  const [hub] = getTestClient((event) => {
    assertSnapshot(t, event);
  });

  function something() {
    return new Error('Some unhandled error');
  }

  hub.captureException(something());
});

Deno.test('captureMessage', (t) => {
  const [hub] = getTestClient((event) => {
    assertSnapshot(t, event);
  });

  hub.captureMessage('Some error message');
});
