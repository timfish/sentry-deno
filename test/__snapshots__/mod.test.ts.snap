export const snapshot = {};

snapshot[`captureException 1`] = `
{
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    deno: {
      name: "Deno",
      target: "{{target}}",
      type: "runtime",
      version: "{{version}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  exception: {
    values: [
      {
        mechanism: {
          handled: true,
          type: "generic",
        },
        stacktrace: {
          frames: [
            {
              colno: 24,
              context_line: "  hub.captureException(something());",
              filename: "app:///test/mod.test.ts",
              function: "<anonymous>",
              in_app: true,
              lineno: 41,
              post_context: [
                "});",
                "",
                "Deno.test('captureMessage', (t) => {",
                "  const [hub] = getTestClient((event) => {",
                "    assertSnapshot(t, event);",
                "  });",
                "",
              ],
              pre_context: [
                "    assertSnapshot(t, event);",
                "  });",
                "",
                "  function something() {",
                "    return new Error('Some unhandled error');",
                "  }",
                "",
              ],
            },
            {
              colno: 12,
              context_line: "    return new Error('Some unhandled error');",
              filename: "app:///test/mod.test.ts",
              function: "something",
              in_app: true,
              lineno: 38,
              post_context: [
                "  }",
                "",
                "  hub.captureException(something());",
                "});",
                "",
                "Deno.test('captureMessage', (t) => {",
                "  const [hub] = getTestClient((event) => {",
              ],
              pre_context: [
                "",
                "Deno.test('captureException', (t) => {",
                "  const [hub] = getTestClient((event) => {",
                "    assertSnapshot(t, event);",
                "  });",
                "",
                "  function something() {",
              ],
            },
          ],
        },
        type: "Error",
        value: "Some unhandled error",
      },
    ],
  },
  sdk: {
    integrations: [
      "Breadcrumbs",
      "Dedupe",
      "GlobalHandlers",
      "ContextLines",
      "TraceFetch",
      "DenoContext",
      "NormalizePaths",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "npm:@sentry/core",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  timestamp: 0,
  user: {
    ip_address: "{{auto}}",
  },
}
`;

snapshot[`captureMessage 1`] = `
{
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    deno: {
      name: "Deno",
      target: "{{target}}",
      type: "runtime",
      version: "{{version}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  level: "info",
  message: "Some error message",
  sdk: {
    integrations: [
      "Breadcrumbs",
      "Dedupe",
      "GlobalHandlers",
      "ContextLines",
      "TraceFetch",
      "DenoContext",
      "NormalizePaths",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "npm:@sentry/core",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  timestamp: 0,
  user: {
    ip_address: "{{auto}}",
  },
}
`;
